tools/scripts/ffi.toml

```diff
@@ -312,4 +313,137 @@ else
+
+# Ensure everything works
+exec --fail-on-error ./bin/gn --version
+exec --fail-on-error ./bin/gnaw --help
+
+echo "*** GN Tools Successfully Installed! Ignore all previous messages. ***"
+'''
+
+[tasks.gn-gen]
+description = "Generate GN build rules for ICU4X"
+category = "ICU4X Development"
+script_runner = "@duckscript"
+script = '''
+exit_on_error true
+
+third_party_tools = is_path_exists ffi/gn/third_party_tools
+assert ${third_party_tools} "The GN third-party tools are not installed.\n*** Please run 'cargo make gn-install' ***"
+
+cd ffi/gn
+exec rm -f icu4x_root
+exec ln -s ../.. icu4x_root
+echo "NOTE: Created a temporary symbolic link at /ffi/gn/icu4x_root"
+
+pwd = pwd
+rm -rf vendor
+output = exec --fail-on-error cargo vendor
+writefile .cargo/config "# This file is part of ICU4X. For terms of use, please see the file\n"
+appendfile .cargo/config "# called LICENSE at the top level of the ICU4X source tree\n"
+appendfile .cargo/config "# (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).\n\n"
+appendfile .cargo/config "# AUTO-GENERATED in ffi.toml\n"
+appendfile .cargo/config ${output.stdout}
+exec --fail-on-error ./third_party_tools/bin/gnaw --manifest-path "${pwd}/Cargo.toml" --project-root ${pwd} --output ${pwd}/icu4x/BUILD.gn --skip-root --gn-bin ${pwd}/third_party_tools/bin/gn
+
+exec rm -f icu4x_root
+echo "NOTE: Removed the temporary symbolic link /ffi/gn/icu4x_root"
+'''
+
+[tasks.verify-gn-gen]
+description = "Verify that checked-in GN rules are up to date"
+category = "ICU4X Development"
+dependencies = [
+    "gn-gen",
+]
+script_runner = "@duckscript"
+script = '''
+exit_on_error true
+code = exec --get-exit-code git diff --exit-code -- ffi/gn
+if ${code}
+    trigger_error "GN rules need to be regenerated. Please run `cargo make gn-install`, then `cargo make gn-gen`, and commit."
+end
+'''
+
+[tasks.gn-build]
+description = "Build the GN version of ICU4X"
+category = "ICU4X Development"
+script_runner = "@duckscript"
+script = '''
+exit_on_error true
+
+third_party_tools = is_path_exists ffi/gn/third_party_tools
+assert ${third_party_tools} "The GN third-party tools are not installed.\n*** Please run 'cargo make gn-install' ***"
+
+cd ffi/gn
+exec rm -f icu4x_root
+exec ln -s ../.. icu4x_root
+echo "NOTE: Created temporary symbolic link at /ffi/gn/icu4x_root"
+
+exec --fail-on-error ./third_party_tools/bin/gn gen out
+exec --fail-on-error rustup run ${ICU4X_NIGHTLY_TOOLCHAIN} ./third_party_tools/depot_tools/ninja -C out
+
+exec rm -f icu4x_root
+echo "NOTE: Removed the temporary symbolic link /ffi/gn/icu4x_root"
+'''
+
+[tasks.gn-run]
+description = "Run the GN version of ICU4X"
+category = "ICU4X Development"
+install_crate = { crate_name = "wasmer", binary = "wasmer", test_arg = ["--help"] }
+dependencies = [ "gn-build" ]
+script_runner = "@duckscript"
+script = '''
+exit_on_error true
+cd ffi/gn
+exec --fail-on-error wasmer run ./out/obj/hello_world/print_hello
```

Comment:
question: wasmer? i thought it was a regular binary? (seems like one? did i miss a line of gn somewhere?)

Reply:
oh right in the BUILDCONFIG.gn. forgot about that, makes sense.

(any reason why not a normal binary?)

Reply:
I thought it was important to test that cross-compiling works.

Reply:
ah fair! should we have a normal test too?

Reply:
Added a normal test.